src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1078,27 +1078,36 @@
                            fact-binding (conj fact-binding))
 
             ;; Find children that all parent nodes have.
-            common-children-ids (into
-                                 ;; Since we require that id-to-condition-nodes have an equal value to "node" under the ID
-                                 ;; for this to be used, the order of this set probably doesn't matter in most cases.  However, in any possible edge
-                                 ;; cases where there are equal nodes under different IDs, sorting the set will add determinism.
-                                 ;; Having different nodes under the same
-                                 ;; ID would be a bug, but having an equivalent node under multiple IDs wouldn't necessarily be one.
-                                 (sorted-set)
-                                 (apply s/union (-> (:forward-edges beta-graph)
-                                                    (select-keys parent-ids)
-                                                    (vals))))
-
+            forward-edges (if (= 1 (count parent-ids))
+                            ;; If there is only one parent then there is no need to reconstruct
+                            ;; the set of forward edges
+                            (-> (:forward-edges beta-graph)
+                                (get (first parent-ids)))
+                            (->> (select-keys parent-ids (:forward-edges beta-graph))
+                                 vals
+                                 (into [] (comp cat (distinct)))))
 
             id-to-condition-nodes (:id-to-condition-node beta-graph)
 
-            ;; Find an id for an equivalent node.
-            existing-id (first (for [child-id common-children-ids
-                                     :when (= node (get id-to-condition-nodes child-id))]
-                                 child-id))
+            ;; Since we require that id-to-condition-nodes have an equal value to "node" under the ID
+            ;; for this to be used, the order of this set probably doesn't matter in most cases.  However, in any possible edge
+            ;; cases where there are equal nodes under different IDs, sorting the set will add determinism.
+            ;; Having different nodes under the same
+            ;; ID would be a bug, but having an equivalent node under multiple IDs wouldn't necessarily be one.
```

Comment:
This comment is out of date now since this isn't a set after these changes - could you remove this and add appropriate comments explaining that we're maintaining deterministic ordering (along with how and why) along with an explanation of the optimizations?